src/notebooks/controllers/kernelSource/kernelSelector.ts

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+'use strict';
+
+import {
+    CancellationToken,
+    CancellationTokenSource,
+    Disposable,
+    Event,
+    NotebookDocument,
+    QuickInputButton,
+    QuickPick,
+    QuickPickItem,
+    QuickPickItemKind,
+    ThemeIcon
+} from 'vscode';
+import { ContributedKernelFinderKind } from '../../../kernels/internalTypes';
```

Comment:
Moved from another file so its easier to maintain & test
In the process found 1 bug, yay for writing tests